File: C(++)/011. Container With Most Water.c
/*
011. Container With Most Water
https://leetcode.com/problems/container-with-most-water/

c 9 ms
*/

/*
Given n non-negative integers a1, a2, ..., an, where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0). Find two lines, which together with x-axis forms a container, such that the container contains the most water.

Note: You may not slant the container and n is at least 2.
*/

int imin(int a,int b)
{
    return a<b?a:b;
}
int imax(int a,int b)
{
    return a>b?a:b;
}
#define VOL (imin(*s,*e)*(ei-si))
int maxArea(int* height, int heightSize) {
    int *s=height,*e=s+heightSize-1,si=1,ei=heightSize,ans=VOL;
    int c=1;
    while (si<ei)
    {
        while (si<ei&&*s<=*e)
        {
            s++;
            si++;
            ans=imax(ans,VOL);
        }
        while (si<ei&&*s>*e)
        {
            e--;
            ei--;
            ans=imax(ans,VOL);
        }
    }
    return ans;
}

File: Java/011. Container With Most Water.java
/*
011. Container With Most Water
https://leetcode.com/problems/container-with-most-water/

java 11 ms
*/

public class Solution {
	public int imin(int a,int b)
	{
		return a<b?a:b;
	}
	public int maxArea(int[] height) {
		int ret=-1,ht[]=height,head=0,tail=ht.length-1;
		/*
		一端h的高度小于另一端t的高度，那么如果选择h端作为一条边，结果就确定了，
		t向左移动只会让结果变小；然而t端的高度没有充分利用，所以可以尝试将h端右移。
		较低的那一端总是被充分利用，所以将较低的一端“移向”较高的一端以寻找更好的结果。
		*/
		while (head<tail)
		{
			int area=(tail-head)*imin(ht[head],ht[tail]);
			if (ret<area) ret=area;
			
			if (ht[head]<ht[tail]) head++;
			else tail--;
		}
		return ret;
	}
}
